### DISCUSSION TITLE
✨ Support Scheduling Build Steps on K8s Nodes

### DISCUSSION THREAD

**Discussion by Jitsusama** [dagger/dagger#12054] [Ideas]
### What are you trying to do?

We have the circumstance where a build could engage a large amount of expensive steps to all be engaged in order to assemble the pieces necessary to create a final build artifact. A number of these steps are large enough that even a well equipped machine might have a hard time coping with it.

My intent is to be able to schedule some/all actions as separate Pods scheduled on a K8s cluster. We could define their resource constraints so that K8s would be able to schedule them to run on nodes that have the needed capacity. This would also prove useful in the case the BuildKit starts supporting Windows Containers (not Linux containers running on Windows via Hyper-V, but actual Windows Containers). We could specify a node tag that would cause steps that require Windows containers to run on Windows K8s nodes &etc.

### Why is this important to you?

Because we have large jobs that couldn't run on a dev machine unless it was super beefy, and even on a server, if run in parallel with other jobs, probably would get too large to run there, this is really a pre-requisite for any build system we would be able to utilize in our industry. We are building FPGAs and embedded software with lots of moving parts making very elaborate builds. We need to be able to run the steps in parallel and be properly scheduled in order to meet our speed and practicality requirements.

### How are you currently working around this?

We have an elaborate bespoke build system in place that has pools of servers able to handle the load, or is able to block builds until resources become available. It still has its issues, which has led us to look into building a new K8s powered build system, but we would prefer to work in tandem with a product like this to pool resources and make this sort of functionality available to the world.

**Comment by TomChv** [comment]
That's an amazing idea, being able to split task into multiple pod.
It requires a lot of thinking and design but result would be awesome!

I'm not sure it's a priority for now, I have tagged this issue with `discussion` so  it will not be forgotten

**Comment by Jitsusama** [comment]
My team would be open to discuss this sometime soon if you'd like. We have a design for a new bespoke system that would fit our needs that would be a large effort for us to build. If we could come to an arrangement perhaps our effort could be spent making this sort of thing work over Dagger instead.

**Comment by Jitsusama** [comment]
https://www.youtube.com/watch?v=wTENRhYt3mw is a video by Apple on front-ending BuildKit with K8s jobs so you can pass through resource constraints and have K8s schedule it. If this could somehow be hooked into Dagger, it would be a possible way to tackle this.
